tests/unit/browser/test_mhtml.py

```diff
@@ -0,0 +1,278 @@
+# vim: ft=python fileencoding=utf-8 sts=4 sw=4 et:
+import io
+import textwrap
+import re
+import pytest
+
+from qutebrowser.browser import mhtml
+
+@pytest.fixture(autouse=True)
+def patch_uuid(monkeypatch):
+    monkeypatch.setattr("uuid.uuid4", lambda: "UUID")
+
+
+class Checker:
+
+    """A helper to check mhtml output.
+
+    Attrs:
+        fp: A BytesIO object for passing to MHTMLWriter.write_to.
+    """
+
+    def __init__(self):
+        self.fp = io.BytesIO()
+
+    @property
+    def value(self):
+        return self.fp.getvalue()
+
+    def expect(self, expected):
+        actual = self.value.decode('ascii')
+        # Make sure there are no stray \r or \n
+        assert re.search(r'\r[^\n]', actual) is None
+        assert re.search(r'[^\r]\n', actual) is None
+        actual = actual.replace('\r\n', '\n')
+        expected = textwrap.dedent(expected).lstrip('\n')
+        assert expected == actual
+
+
+@pytest.fixture
+def checker():
+    return Checker()
+
+
+def test_quoted_printable_umlauts(checker):
+    content = 'Die süße Hündin läuft in die Höhle des Bären'
+    content = content.encode('iso-8859-1')
+    writer = mhtml.MHTMLWriter(root_content=content,
+                               content_location='localhost',
+                               content_type='text/plain')
+    writer.write_to(checker.fp)
+    checker.expect("""
+        Content-Type: multipart/related; boundary="---=_qute-UUID"
+        MIME-Version: 1.0
+
+        -----=_qute-UUID
+        Content-Location: localhost
+        MIME-Version: 1.0
+        Content-Type: text/plain
+        Content-Transfer-Encoding: quoted-printable
+
+        Die=20s=FC=DFe=20H=FCndin=20l=E4uft=20in=20die=20H=F6hle=20des=20B=E4ren
+        -----=_qute-UUID--
+        """)
+
+
+@pytest.mark.parametrize('header, value', [
+    ('content_location', 'http://brötli.com'),
+    ('content_type', 'text/pläin'),
+])
+def test_refuses_non_ascii_header_value(checker, header, value):
+    defaults = {
+        'root_content': b'',
+        'content_location': 'http://example.com',
+        'content_type': 'text/plain',
+    }
+    defaults[header] = value
+    writer = mhtml.MHTMLWriter(**defaults)
+    with pytest.raises(UnicodeEncodeError) as excinfo:
+        writer.write_to(checker.fp)
+    assert "'ascii' codec can't encode" in str(excinfo.value)
```

Comment:
I tested this [using httpbin](http://httpbin.org/response-headers?Content-Type=text%2Fpl%C3%A4in) and I get:

```
Traceback (most recent call last):
  File "/home/florian/proj/qutebrowser/git/qutebrowser/browser/mhtml.py", line 443, in start_download_checked
    start_download(dest)
  File "/home/florian/proj/qutebrowser/git/qutebrowser/browser/mhtml.py", line 429, in start_download
    loader.run()
  File "/home/florian/proj/qutebrowser/git/qutebrowser/browser/mhtml.py", line 282, in run
    self.finish_file()
  File "/home/florian/proj/qutebrowser/git/qutebrowser/browser/mhtml.py", line 385, in finish_file
    self.writer.write_to(file_output)
  File "/home/florian/proj/qutebrowser/git/qutebrowser/browser/mhtml.py", line 180, in write_to
    gen.flatten(msg)
  File "/usr/lib/python3.4/email/generator.py", line 112, in flatten
    self._write(msg)
  File "/usr/lib/python3.4/email/generator.py", line 178, in _write
    self._dispatch(msg)
  File "/usr/lib/python3.4/email/generator.py", line 211, in _dispatch
    meth(msg)
  File "/usr/lib/python3.4/email/generator.py", line 269, in _handle_multipart
    g.flatten(part, unixfrom=False, linesep=self._NL)
  File "/usr/lib/python3.4/email/generator.py", line 112, in flatten
    self._write(msg)
  File "/usr/lib/python3.4/email/generator.py", line 192, in _write
    self._write_headers(msg)
  File "/usr/lib/python3.4/email/generator.py", line 419, in _write_headers
    self._fp.write(self.policy.fold_binary(h, v))
  File "/usr/lib/python3.4/email/policy.py", line 180, in fold_binary
    return folded.encode('ascii', 'surrogateescape')
UnicodeEncodeError: 'ascii' codec can't encode character '\xe4' in position 76: ordinal not in range(128)
```

So yeah, would be great to handle this in some sane way - let's hope it never happens, but I'd still rather not have a crash if it does :wink: